nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java

```diff
@@ -107,110 +174,200 @@ public NullnessHint onDataflowVisitMethodInvocation(
+                  .createErrorDescription(
+                      new ErrorMessage(
+                          ErrorMessage.MessageTypes.ANNOTATION_VALUE_INVALID, errorMessage),
+                      tree,
+                      analysis.buildDescription(tree),
+                      state,
+                      null));
+          supported = false;
+          break;
         }
-        if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULLABLE)) {
-          // If argIdx being null implies the return of the method being false, then the return
-          // being true implies argIdx is not null and we must mark it as such in the then update.
-          thenUpdates.set(accessPath, Nullness.NONNULL);
-        } else if (consequent.equals("true") && argAntecedentNullness.equals(Nullness.NULLABLE)) {
-          // If argIdx being null implies the return of the method being true, then the return being
-          // false implies argIdx is not null and we must mark it as such in the else update.
-          elseUpdates.set(accessPath, Nullness.NONNULL);
-        } else if (consequent.equals("fail") && argAntecedentNullness.equals(Nullness.NULLABLE)) {
-          // If argIdx being null implies the method throws an exception, then we can mark it as
-          // non-null on both non-exceptional exits from the method
-          bothUpdates.set(accessPath, Nullness.NONNULL);
+      }
+      if (!supported) {
+        // Too many arguments involved, or unsupported @Contract features. On to next clause in
+        // the
+        // contract expression
+        continue;
+      }
+      if (arg == null) {
+        // The antecedent is unconditionally true. Check for the ... -> !null case and set the
+        // return nullness accordingly
+        if (consequent.equals("!null")) {
+          return NullnessHint.FORCE_NONNULL;
         }
+        continue;
+      }
+      if (argAntecedentNullness == null) {
+        throw new IllegalStateException("argAntecedentNullness should have been set");
+      }
+      // The nullness of one argument is all that matters for the antecedent, let's negate the
+      // consequent to fix the nullness of this argument.
+      AccessPath accessPath = AccessPath.getAccessPathForNodeNoMapGet(arg, apContext);
+      if (accessPath == null) {
+        continue;
+      }
+      if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NULL)) {
+        // If arg being null implies the return of the method being false, then the return
+        // being true implies arg is not null and we must mark it as such in the then update.
+        thenUpdates.set(accessPath, Nullness.NONNULL);
+      } else if (consequent.equals("false") && argAntecedentNullness.equals(Nullness.NONNULL)) {
+        // If arg being non-null implies the return of the method being false, then the return
+        // being true implies arg may be null and we must mark it as such in the then update.
+        thenUpdates.set(
```

Comment:
These cases where we set access paths to `NULLABLE.greatestLowerBound(...)` don't impact the final NullAway results, right?  I.e., they never cause NullAway to report additional bugs?  If that's true I'd prefer to skip these access, to avoid tracking unnecessary access paths in the store (for performance)

Reply:
A test case here would be great; I still don't immediately see how to write one.

Reply:
I've added relevant test coverage for both `!null -> false` and boolean-based equivalent. It turned out the former was already covered by this condition above: https://github.com/uber/NullAway/blob/4136de7a12eb131020d65811df918859e9209339/nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractHandler.java#L137-L142

Given that, I've implemented a similar check for the boolean path and removed the `greatestLowerBound` calls which are no longer necessary.